state/state.go

```diff
@@ -119,21 +143,62 @@ func (s *State) LoadABCIResponses() *ABCIResponses {
 	return abciResponses
 }
 
+// SaveValidators persists the validator set to disk. If the validator set did not change in the latest block,
+// only the last height in which the validators changed is persisted.
+func (s *State) SaveValidators() {
+	lastHeight := s.lastHeightValidatorsChanged
+	thisHeight := s.LastBlockHeight
+	v := &Validators{
+		LastHeightChanged: lastHeight,
+	}
+	if lastHeight == thisHeight {
+		v.ValidatorSet = s.Validators
+	}
+	s.db.SetSync(calcValidatorsKey(thisHeight), v.Bytes())
```

Comment:
It will either succeed or Panic: https://github.com/tendermint/tmlibs/blob/master/db/go_level_db.go#L58